exonum/src/runtime/rust/call_context.rs

```diff
@@ -0,0 +1,200 @@
+use exonum_merkledb::{BinaryValue, Fork};
+
+use super::super::{
+    dispatcher::{Dispatcher, Error as DispatcherError, Schema},
+    error::catch_panic,
+    ArtifactId, CallInfo, Caller, ExecutionContext, ExecutionError, InstanceDescriptor, InstanceId,
+    InstanceQuery, InstanceSpec, MethodId, SUPERVISOR_INSTANCE_ID,
+};
+
+/// Context for the executed call.
+///
+/// The call can mean a transaction call, or the `before_commit` hook.
+#[derive(Debug)]
+pub struct CallContext<'a> {
+    /// Underlying execution context.
+    inner: ExecutionContext<'a>,
+    /// ID of the executing service.
+    instance: InstanceDescriptor<'a>,
+}
+
+impl<'a> CallContext<'a> {
+    /// Creates a new transaction context for the specified execution context and the instance
+    /// descriptor.
+    pub(crate) fn new(context: ExecutionContext<'a>, instance: InstanceDescriptor<'a>) -> Self {
+        Self {
+            inner: context,
+            instance,
+        }
+    }
+
+    /// Returns a writable snapshot of the current blockchain state.
+    pub fn fork(&self) -> &Fork {
+        self.inner.fork
+    }
+
+    /// Returns the initiator of the actual transaction execution.
+    pub fn caller(&self) -> &Caller {
+        &self.inner.caller
+    }
+
+    /// Accesses dispatcher information.
+    pub fn dispatcher(&self) -> Schema<&Fork> {
+        Schema::new(self.fork())
```

Comment:
Tried `self.fork().as_ref()`; it doesn't work because `Fork::as_ref()` does not capture non-flushed changes in the fork. I think this is a problem with MerkleDB APIs, specifically, with lack of split between read-only methods and full-access methods. I believe we've already discussed this design flaw.

> no service will be able to screw the dispatcher schema up?

It will be possible as long as the schema constructor is public, *or* if there is no split of read-only / full access methods of the indexes / schemas as described above. It could make sense to restrict write access to the schema of the service being executed to alleviate similar concerns once and for all, but it is a separate task.